pandas/core/arrays/categorical.py

```diff
@@ -2439,11 +2439,16 @@ def _get_codes_for_values(values, categories):
     """
     utility routine to turn values into codes given the specified categories
     """
-
     from pandas.core.algorithms import _get_data_algo, _hashtables
     if not is_dtype_equal(values.dtype, categories.dtype):
         values = ensure_object(values)
         categories = ensure_object(categories)
+    else:
+        # These may be Index, in which case their dtype would be coerced
+        # as part of _get_data_algo; in that case, we should retrieve the
+        # underlying numpy array instead.
+        values = getattr(values, 'values', values)
```

Comment:
I think you can just use ``np.asarray`` here as we have to re-infer anyhow. But pls check perf of this. The comment can be a bit simpler, and reference the gh issue number

Reply:
```python
In [4]: categories = pd.Index([True, False], dtype='object')

In [5]: values = np.array([], dtype='object')

In [6]: %timeit np.asarray(values)
260 ns ± 1.41 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [7]: %timeit np.asarray(categories)
2.2 µs ± 21.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit getattr(values, 'values', values)
312 ns ± 1.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit getattr(categories, 'values', categories)
410 ns ± 4.23 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [10]: categories = pd.Index(list('abcdefghijklmnopqrstuvwxyz'), dtype='object')

In [11]: categories
Out[11]: 
Index(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n',
       'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'],
      dtype='object')

In [12]: %timeit np.asarray(categories)
2.09 µs ± 9.76 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [13]: %timeit getattr(categories, 'values', categories)
409 ns ± 6.44 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each
```

Looks like the way I've done it is faster, but `np.asarray` doesn't seem to be scaling with the number of categories, so I don't think it's a huge deal to use `np.asarray`. That said, I don't love the idea of using `np.asarray` in this situation, because although I can't *think* of a time when it would present a problem, `np.asarray` is allowed to infer `dtype`. The way I've done it you're guaranteed to either get the original thing back or the underlying numpy array, which is pretty much as light a touch as I can imagine.